src/search/tasks/negated_axioms_task.cc

```diff
@@ -0,0 +1,358 @@
+#include "negated_axioms_task.h"
+
+#include "../operator_cost.h"
+
+#include "../plugins/plugin.h"
+#include "../task_utils/task_properties.h"
+#include "../tasks/root_task.h"
+
+#include <deque>
+#include <iostream>
+#include <memory>
+#include <set>
+
+using namespace std;
+using utils::ExitCode;
+
+namespace tasks {
+NegatedAxiomsTask::NegatedAxiomsTask(const shared_ptr<AbstractTask> &parent)
+    : DelegatingTask(parent),
+      negated_axioms_start_index(parent->get_num_axioms()) {
+    TaskProxy task_proxy(*parent);
+
+    vector<vector<int>> pos_dependencies(task_proxy.get_variables().size());
+    vector<vector<int>> neg_dependencies(task_proxy.get_variables().size());
+    vector<vector<int>> axiom_ids_for_var(task_proxy.get_variables().size());
+    for (OperatorProxy axiom: task_proxy.get_axioms()) {
+        EffectProxy effect = axiom.get_effects()[0];
+        int head_var = effect.get_fact().get_variable().get_id();
+        axiom_ids_for_var[head_var].push_back(axiom.get_id());
+        for (FactProxy condition: effect.get_conditions()) {
+            VariableProxy var_proxy = condition.get_variable();
+            if (var_proxy.is_derived()) {
+                int var = condition.get_variable().get_id();
+                if (condition.get_value() == var_proxy.get_default_axiom_value()) {
+                    neg_dependencies[head_var].push_back(var);
+                } else {
+                    pos_dependencies[head_var].push_back(var);
+                }
+            }
+        }
+    }
+    unordered_set<int> needed_negatively =
+        collect_needed_negatively(pos_dependencies, neg_dependencies);
+
+    /* TODO: I removed the return if needed_negatively is empty.
+       It is somewhat wasteful if we have a big graph to compute the sccs,
+       but probably should not be an issue. */
+
+    /*
+       Get the sccs induced by positive dependencies.
+       Note that negative dependencies cannot introduce additional
+       cycles because this would imply that the axioms are not
+       stratifiable, which is already checked in the translator.
+    */
+    vector<vector<int>> sccs =
+        sccs::compute_maximal_sccs(pos_dependencies);
+    vector<int> var_to_scc(task_proxy.get_variables().size(), -1);
+    for (int i = 0; i < (int)sccs.size(); ++i) {
+        for (int var: sccs[i]) {
+            var_to_scc[var] = i;
+        }
+    }
```

Comment:
Maybe I am coming too much from the Python side here but wouldn't it be more natural to store in `var_to_scc` references to the respective SCCs rather than an index into `sccs`?

Reply:
References in vectors are very cumbersome, but i used pointers now.